clang/bindings/python/clang/cindex.py

```diff
@@ -386,6 +386,10 @@ def __contains__(self, other):
         # same file, in between lines
         if self.start.line < other.line < self.end.line:
             return True
+        # between columns in one-liner range
+        elif self.start.line == other.line == self.end.line:
```

Comment:
How about we follow the `SourceRange::fullyContains`? In other words, expose `SourceLocation::operator<` to enable ordering of source locations, and leverage that in `SourceRange.__contains__`. I expect the latter to become much simpler.

Reply:
I said "expose", but it'd be better just reimplement it in Python than actually export it from C++ code.

Reply:
I was just about to push a  new `clang_locationLeq` funcion :D
Great idea though, implemented `SourceLocation.__le__ `

Reply:
Is there a reason you implemented _less-or-equals_ operator, and not _less_? `operator<` + `operator==` is sort of a canonical pair of operators to establish ordering, and it would match `SourceLocation` in C++.

Reply:
`SourceRange::fullyContains` implements the check via `operator<=`, which is the most straightforward way, and since the intention is not really to expand the interface but we're just adding the operator to simplify `SourceRange.__contains__` I went with this. Happy to remove this and instead implement `SourceLocation.__lt__` though if you prefer.

That said, the situation is a bit ugly though: we already have `SourceLocation.__eq__` implemented (which just passes directly to the C interface) and that one _is file-sensitive_. I.e. `loc1 == loc2` gives false if their line & column are the same but their file name is different, while `loc1 <= loc2` is `True` in that case. This is already the case in the C interface.
So we can't really use `SourceLocation.__eq__` in `SourceRange.__contains__` because the latter wants to check only if their lines/columns match, disregarding filename, so it works with half-open ranges where there is no end `SourceLocation`.

Reply:
> we already have SourceLocation.__eq__ implemented (which just passes directly to the C interface)

I didn't realize that, thank you!

> and that one is file-sensitive. I.e. loc1 == loc2 gives false if their line & column are the same but their file name is different, while loc1 <= loc2 is True in that case.

This is sure ugly.

I see 4 ways forward:
1) Carry on with implementing `__contains__` in Python in terms of line and column numbers. I don't like that precisely because of its file-insensitivity.
2) Do what `clang_equalLocations` does, i.e. add a libclang function that reimplements `SourceLocation::operator<` in terms of `CXSourceLocation.int_data` (which corresponds to `SourceLocation::ID` the original `operator<` is implemented in terms of).
3) Do what `clang_equalLocations` does, but in Python, i.e. implement `SourceLocation.__contains__` in terms of `SourceLocation.int_data`.
4) Literally expose `SourceLocation::operator<` via libclang as I suggested earlier.

I don't like option 1, because we'd intentionally provide a broken API. I cautiously support option 3 as having minimal impact on various interfaces, but I'm perplexed that libclang interprets guts of `SourceLocation` (which in turn is an opaque value that only `SourceManager` can make sense of).

Now I wonder how well libclang is tested w.r.t. potential changes in the underlying C++ API. If I were to change something in `SourceLocation` layout, libclang would be the last thing on my mind, even though it's our stable API (and ABI, I believe).

CC @AaronBallman I wonder what your thoughts are

Reply:
Clarification because I probably expressed myself inaccurately:
On C-side, in `SourceLocation.h` we implement `operator==`, `operator<=` both as direct comparisons on the raw encoding (`SourceLocation::getRawEncoding()`). This is not file-sensitive.*
On the C/Python _bindings_ side, we already expose `clang_equalLocations`, which looks like this:
```
unsigned clang_equalLocations(CXSourceLocation loc1, CXSourceLocation loc2) {
  return (loc1.ptr_data[0] == loc2.ptr_data[0] &&
          loc1.ptr_data[1] == loc2.ptr_data[1] &&
          loc1.int_data == loc2.int_data);
}
```
This _is_ file sensitive. The problem arises because this function is used already to implement `SourceLocation.__eq__` on the Python bindings side. This means that the internal `SourceLocation::operator==` is inconsistent with the bindings' `SourceLocation.__eq__` (the latter is file-sensitive, the former is not). The C-bindings are not affected by this as they do not implement `CXSourceLocation::opreraor==` or `CXSourceLocation::opreraor<=`.


*Note that this is not obvious from the code (at least to me) and I'm not familiar with the C side yet, so I tested this manually in Python, by exposing these operators to the Python bindings (pls don't judge me)

Reply:
> On C++ side, in SourceLocation.h we implement operator==, operator<= both as direct comparisons on the raw encoding (SourceLocation::getRawEncoding()). This is not file-sensitive.*

I believe raw encoding is file-sensitive.

If you say that `clang_equalLocations` is file-sensitive, then `SourceLocation::operator==` is file-sensitive, too. If you look at how `CXSourceLocation` are initialized, `int_data` is raw encoding, and `ptr_data[0]` is `SourceManager` (`ptr_data[1]` seems to be a dead weight).

Reply:
As I understand it, raw encoding can be thought of as an offset in preprocessed source where `#include`s and macros are substituted, but macro definitions are kept intact for source location purposes.

Reply:
Okay, that was my mistake: when testing this, I created two separate TUs for the two files, and _that_ is apparently where it's sensitive. So no issue on the C-side here after al.

Going back to the Python bindings: we are now left with a pre-existing `SourceLocation.__eq__` that _is_ file-sensitive, and in order to have a clean implementation, we would like a `SourceLocation.__le__` / `SourceLOcation.__lt__` that is _not_ file-sensitive, so it can be used for `SourceRange.__contains__`. 
Exposing something like this would be ugly due to inconsistency. So I guess it would be the easiest to implement such an `operator<=` function just locally inside `SourceRange.__contains__` to simplify the logic there, and not expose any new functionality at all. What do you think?

Reply:
> and in order to have a clean implementation, we would like a SourceLocation.__le__ / SourceLOcation.__lt__ that is not file-sensitive

I'm against anything non file-sensitive, because that's buggy and is going to surprise everyone. I'm less concerned with cleanliness of implementation in the face of this concern.

At this point I'm even more concerned by amount of places that interpret raw encoding of `SourceLocation`, even though (in my understanding) `SourceManager` is the only entity that should do so. I'd like to hear from Aaron on Monday on this. After that it should be more clear how we should proceed on all three levels of our API.

Reply:
> I'm against anything non file-sensitive, because that's buggy and is going to surprise everyone. I'm less concerned with cleanliness of implementation in the face of this concern.

+1, that would be surprising behavior.

> At this point I'm even more concerned by amount of places that interpret raw encoding of SourceLocation, even though (in my understanding) SourceManager is the only entity that should do so.

The raw encoding of source locations is an implementation detail and should not be relied on. You don't always have to go to the `SourceManager` (for example, you can get the raw encoding from one `SourceLocation` and use it to construct an equivalent `SourceLocation` through the correct constructor), but we should not be relying on the representation to be stable.

It should also be noted that modules may break everything that's comparing raw locations:
```
/// Technically, a source location is simply an offset into the manager's view
/// of the input source, which is all input buffers (including macro
/// expansions) concatenated in an effectively arbitrary order. The manager
/// actually maintains two blocks of input buffers. One, starting at offset
/// 0 and growing upwards, contains all buffers from this module. The other,
/// starting at the highest possible offset and growing downwards, contains
/// buffers of loaded modules.
```

Reply:
I've exposed a `clang_lessThanLocations` function on C-side, similar to the `clang_equalLocations` function, and used this to implement the operators on Python side. Hope that resolves this case.